server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -314,30 +278,103 @@ private boolean buildBufferAndMatchRight() {
     }
 
     private void addToBuffer(Object[] currentRow, int hash) {
-        Values existingRows = buffer.get(hash);
-        if (existingRows == null) {
-            existingRows = new Values();
-            buffer.put(hash, existingRows);
+        HashGroup hashGroup = buffer.get(hash);
+        if (hashGroup == null) {
+            hashGroup = new HashGroup(emitUnmatchedRows);
+            buffer.put(hash, hashGroup);
         }
-        existingRows.items.add(currentRow);
-        numberOfRowsInBuffer++;
+        hashGroup.add(currentRow);
+        numberOfHashGroupsInBuffer++;
     }
 
     private boolean findMatchingRows() {
-        while (leftMatchingRowsIterator.hasNext()) {
-            leftRow.cells(leftMatchingRowsIterator.next());
+        List<Object[]> matchedRows = matchedHashGroup.rows;
+        BitSet rowIsJoinedFlags = matchedHashGroup.rowIsJoinedFlags;
+        while (matchedHashGroupIdx < matchedRows.size()) {
+            leftRow.cells(matchedRows.get(matchedHashGroupIdx));
             combiner.setLeft(leftRow);
             if (joinCondition.test(combiner.currentElement())) {
+                if (emitUnmatchedRows) {
+                    rowIsJoinedFlags.set(matchedHashGroupIdx, true);
+                }
+                matchedHashGroupIdx++;
                 return true;
             }
+            matchedHashGroupIdx++;
         }
+        matchedHashGroup = null;
+        matchedHashGroupIdx = 0;
         return false;
     }
 
-    private static final class Values {
+    private static final class HashGroup {
+
+        private final List<Object[]> rows = new ArrayList<>();
+        @Nullable
+        private final BitSet rowIsJoinedFlags;
+
+        private HashGroup(boolean emitUnmatchedRows) {
+            this.rowIsJoinedFlags = emitUnmatchedRows ? new BitSet() : null;
+        }
+
+        public void add(Object[] row) {
+            rows.add(row);
+        }
+    }
+
+    private static class UnmatchedRowsIterator implements Iterator<Object[]> {
+
+        private final Iterator<IntObjectHashMap.PrimitiveEntry<HashJoinBatchIterator.HashGroup>> hashGroupsIterator;
 
-        ArrayList<Object[]> items = new ArrayList<>();
-        boolean matched = false;
+        private HashGroup currentHashGroup;
+        private int currentHashGroupRowsIdx;
+        private Object[] nextUnmatchedRow;
 
+        public UnmatchedRowsIterator(IntObjectHashMap<HashGroup> buffer) {
+            this.hashGroupsIterator = buffer.entries().iterator();
+            advancedToNextUnmatchedRow();
```

Comment:
Could we let `HashGroup` implement an iterator instead and then just concat the different HashGroup.iterators here? I think this would let us implement this iterator methods in an more clean way, maybe even avoiding having states inside this iterator.
This may not change memory allocation or performance, but may result in cleaner and safer code.